File: packages/mirascript/src/vm/operations/operator.ts
import { toFormat } from '../../helpers/convert/index.js';
import { isNaN } from '../../helpers/utils.js';
import type { VmAny } from '../types/index.js';
import { $AssertInit } from './common.js';
import { $ToBoolean, $ToNumber, $ToString } from './convert.js';
import { isSame, overloadNumberString } from './utils.js';

// String operations
/** 字符串连接 */
export function $Concat(...args: readonly VmAny[]): string {
    return args.map((a) => toFormat(a, null)).join('');
}

// Unary operations
/** 正号 */
export const $Pos = $ToNumber;
/** 负号 */
export const $Neg = (a: VmAny): number => -$ToNumber(a);
/** 非 */
export const $Not = (a: VmAny): boolean => !$ToBoolean(a);

// Math operations
const add = (a: VmAny, b: VmAny): number => $ToNumber(a) + $ToNumber(b);
/** 加法 */
export const $Add = (a: VmAny, b: VmAny): number => {
    if (typeof a == 'number' && typeof b == 'number') return a + b;
    return add(a, b);
};
const sub = (a: VmAny, b: VmAny): number => $ToNumber(a) - $ToNumber(b);
/** 减法 */
export const $Sub = (a: VmAny, b: VmAny): number => {
    if (typeof a == 'number' && typeof b == 'number') return a - b;
    return sub(a, b);
};
const mul = (a: VmAny, b: VmAny): number => $ToNumber(a) * $ToNumber(b);
/** 乘法 */
export const $Mul = (a: VmAny, b: VmAny): number => {
    if (typeof a == 'number' && typeof b == 'number') return a * b;
    return mul(a, b);
};
const div = (a: VmAny, b: VmAny): number => $ToNumber(a) / $ToNumber(b);
/** 除法 */
export const $Div = (a: VmAny, b: VmAny): number => {
    if (typeof a == 'number' && typeof b == 'number') return a / b;
    return div(a, b);
};
const mod = (a: VmAny, b: VmAny): number => $ToNumber(a) % $ToNumber(b);
/** 取模 */
export const $Mod = (a: VmAny, b: VmAny): number => {
    if (typeof a == 'number' && typeof b == 'number') return a % b;
    return mod(a, b);
};
const pow = (a: VmAny, b: VmAny): number => $ToNumber(a) ** $ToNumber(b);
/** 乘方 */
export const $Pow = (a: VmAny, b: VmAny): number => {
    if (typeof a == 'number' && typeof b == 'number') return a ** b;
    return pow(a, b);
};

// Logical operations without short-circuiting
const and = (a: VmAny, b: VmAny): boolean => $ToBoolean(a) && $ToBoolean(b);
/** 与 */
export const $And = (a: VmAny, b: VmAny): boolean => {
    if (typeof a == 'boolean' && typeof b == 'boolean') return a && b;
    return and(a, b);
};
const or = (a: VmAny, b: VmAny): boolean => $ToBoolean(a) || $ToBoolean(b);
/** 或 */
export const $Or = (a: VmAny, b: VmAny): boolean => {
    if (typeof a == 'boolean' && typeof b == 'boolean') return a || b;
    return or(a, b);
};

// Comparison operations
/** 大于 */
export const $Gt = (a: VmAny, b: VmAny): boolean => {
    if (overloadNumberString(a, b)) {
        return $ToNumber(a) > $ToNumber(b);
    } else {
        return $ToString(a) > $ToString(b);
    }
};
/** 大于等于 */
export const $Gte = (a: VmAny, b: VmAny): boolean => {
    if (overloadNumberString(a, b)) {
        return $ToNumber(a) >= $ToNumber(b);
    } else {
        return $ToString(a) >= $ToString(b);
    }
};
/** 小于 */
export const $Lt = (a: VmAny, b: VmAny): boolean => {
    if (overloadNumberString(a, b)) {
        return $ToNumber(a) < $ToNumber(b);
    } else {
        return $ToString(a) < $ToString(b);
    }
};
/** 小于等于 */
export const $Lte = (a: VmAny, b: VmAny): boolean => {
    if (overloadNumberString(a, b)) {
        return $ToNumber(a) <= $ToNumber(b);
    } else {
        return $ToString(a) <= $ToString(b);
    }
};

// Equality operations
/** 等于 */
export const $Eq = (a: VmAny, b: VmAny): boolean => {
    $AssertInit(a);
    $AssertInit(b);
    // Number comparison is a special case to handle NaN correctly
    if (typeof a == 'number' && typeof b == 'number') return a === b;
    return isSame(a, b);
};
/** 不等于 */
export const $Neq = (a: VmAny, b: VmAny): boolean => {
    return !$Eq(a, b);
};

const { abs, min } = Math;
/** 近似等于 */
export const $Aeq = (a: VmAny, b: VmAny): boolean => {
    if (overloadNumberString(a, b)) {
        const an = $ToNumber(a);
        const bn = $ToNumber(b);
        const EPS = 1e-15;
        if (isNaN(an) || isNaN(bn)) return false;
        // Since Inf - Inf is NaN, we must check for equality first
        if (an === bn) return true;
        const absoluteDifference = abs(an - bn);
        if (absoluteDifference < EPS) return true;
        const base = min(abs(an), abs(bn));
        return absoluteDifference < base * EPS;
    } else {
        // For strings, we use normalized case-insensitive comparison
        const as = $ToString(a);
        const bs = $ToString(b);
        if (as === bs) return true;
        const ai = as.toLowerCase();
        const bi = bs.toLowerCase();
        if (ai === bi) return true;
        const an = ai.normalize('NFC');
        const bn = bi.normalize('NFC');
        return an === bn;
    }
};
/** 不近似等于 */
export const $Naeq = (a: VmAny, b: VmAny): boolean => {
    return !$Aeq(a, b);
};

/** 全等于 */
export const $Same = (a: VmAny, b: VmAny): boolean => {
    $AssertInit(a);
    $AssertInit(b);
    return isSame(a, b);
};
/** 不全等于 */
export const $Nsame = (a: VmAny, b: VmAny): boolean => {
    return !$Same(a, b);
};

File: crates/python/mirascript/_vm/operations/operator.py
import math
import unicodedata

from ..._helpers.convert.to_format import toFormat
from .common import AssertInit
from .convert import ToNumber, ToBoolean, ToString
from .utils import is_number, is_same, overload_number_string

# String operations


def Concat(*args) -> str:
    result = ""
    for a in args:
        AssertInit(a)
        result += toFormat(a, None)
    return result


# Unary operations


def Pos(a) -> float:
    return ToNumber(a)


def Neg(a) -> float:
    return -ToNumber(a)


def Not(a) -> bool:
    return not ToBoolean(a)


# Math operations


def Add(a, b) -> float:
    return ToNumber(a) + ToNumber(b)


def Sub(a, b) -> float:
    return ToNumber(a) - ToNumber(b)


def Mul(a, b) -> float:
    return ToNumber(a) * ToNumber(b)


def Div(a, b) -> float:
    a = ToNumber(a)
    b = ToNumber(b)
    if b != 0:
        return a / b

    if a == 0:
        return math.nan
    return math.copysign(1, a) * math.copysign(1, b) * math.inf


def Mod(a, b) -> float:
    x = ToNumber(a)
    y = ToNumber(b)
    """
    IEEE 754 标准的取余运算 (remainder operation)
    对应 ECMAScript 的 % 运算符
    
    规范定义：
    1. If either operand is NaN, the result is NaN.
    2. If x is ±Infinity, the result is NaN.
    3. If y is ±Infinity and x is finite, the result is x.
    4. If y is ±0, the result is NaN.
    5. If x is ±0 and y is nonzero and finite, the result is x.
    6. Otherwise, the result has the same sign as x and magnitude:
       abs(x) - floor(abs(x) / abs(y)) × abs(y)
    """
    # 1. If either operand is NaN, return NaN
    if math.isnan(x) or math.isnan(y):
        return math.nan

    # 2. If x is ±Infinity, return NaN
    if math.isinf(x):
        return math.nan

    # 3. If y is ±Infinity and x is finite, return x
    if math.isinf(y):
        return x

    # 4. If y is ±0, return NaN
    if y == 0:
        return math.nan

    # 5. If x is ±0 and y is nonzero and finite, return x (保留符号)
    if x == 0:
        return x  # 保留 +0 或 -0 的符号

    # 6. 计算取余：结果符号与 x 相同
    # Python 的 % 运算符结果符号与除数相同，不符合 IEEE 754
    # 需要使用 math.fmod，它的结果符号与被除数相同
    return math.fmod(x, y)


def Pow(a, b) -> float:
    try:
        return math.pow(ToNumber(a), ToNumber(b))
    except ValueError:
        return math.nan


# Logical operations without short-circuiting


def And(a, b) -> bool:
    return ToBoolean(a) and ToBoolean(b)


def Or(a, b) -> bool:
    return ToBoolean(a) or ToBoolean(b)


# Comparison operations


def Gt(a, b) -> bool:
    if overload_number_string(a, b):
        return ToNumber(a) > ToNumber(b)
    else:
        return str(a) > str(b)


def Gte(a, b) -> bool:
    if overload_number_string(a, b):
        return ToNumber(a) >= ToNumber(b)
    else:
        return str(a) >= str(b)


def Lt(a, b) -> bool:
    if overload_number_string(a, b):
        return ToNumber(a) < ToNumber(b)
    else:
        return str(a) < str(b)


def Lte(a, b) -> bool:
    if overload_number_string(a, b):
        return ToNumber(a) <= ToNumber(b)
    else:
        return str(a) <= str(b)


def Eq(a, b) -> bool:
    AssertInit(a)
    AssertInit(b)

    if is_number(a) and is_number(b):
        return a == b

    return is_same(a, b)


def Neq(a, b) -> bool:
    return not Eq(a, b)


def Aeq(a, b) -> bool:
    if overload_number_string(a, b):
        a_n = ToNumber(a)
        b_n = ToNumber(b)
        EPS = 1e-15
        if math.isnan(a_n) or math.isnan(b_n):
            return False
        if a_n == b_n:
            return True
        absoluteDifference = abs(a_n - b_n)
        if absoluteDifference < EPS:
            return True
        base = min(abs(a_n), abs(b_n))
        return absoluteDifference < base * EPS
    else:
        a_s = ToString(a)
        b_s = ToString(b)

        if a_s == b_s:
            return True

        a_i = a_s.lower()
        b_i = b_s.lower()

        if a_i == b_i:
            return True

        ai = unicodedata.normalize("NFC", a_i)
        bi = unicodedata.normalize("NFC", b_i)
        return ai == bi


def Naeq(a, b) -> bool:
    return not Aeq(a, b)


def Same(a, b) -> bool:
    AssertInit(a)
    AssertInit(b)
    return is_same(a, b)


def Nsame(a, b) -> bool:
    return not Same(a, b)
